modal/volume.py

```diff
@@ -268,61 +277,164 @@ async def remove_file(self, path: Union[str, bytes], recursive: bool = False) ->
         await retry_transient_errors(self._client.stub.VolumeRemoveFile, req)
 
     @live_method
-    async def _add_local_file(
-        self, local_path: Union[Path, str], remote_path: Optional[Union[str, PurePosixPath, None]] = None
-    ):
-        mount_file = await self._upload_file(local_path, remote_path)
-        request = api_pb2.VolumePutFilesRequest(volume_id=self.object_id, files=[mount_file])
-        await retry_transient_errors(self._client.stub.VolumePutFiles, request, base_delay=1)
+    async def put(self) -> "_VolumeUploadContextManager":
+        """
+        Initiate a batched upload to a volume.
 
-    @live_method
-    async def _add_local_dir(
+        **Example:**
+
+        ```python notest
+        vol = modal.Volume.lookup("my-modal-volume")
+
+        with vol.put() as batch:
+            batch.add_local_file("file-to-upload.txt")  # Will added as /file-to-upload.txt
+            batch.add_local_file("a.txt", "/optional-destination-path.txt")
+            batch.add_local_dir("/some/dir/")  # Will be added as /dir
```

Comment:
I don't think we should support the one-argument API here — it's a little bit confusing why you would add something to the root directory? More clear I think to only have the two-argument API.

```
batch.add_local_file("file-to-upload.txt", "/file-to-upload.txt")
batch.add_local_file("a.txt", "/optional-destination-path.txt")
batch.add_local_dir("/some/dir", "/dir")
```

Is there a reason why `/some/dir/` needs a trailing slash? I think it's clear enough with the `add_local_dir` method name. Trailing slashes have some [confusing semantics](https://tookmund.com/2022/04/importance-of-the-trailing-slash) in `cp` that I always need to look up

Reply:
I don't have strong opinions about the one-argument API. My only reason for having it was to stay consistent with NFS: https://github.com/modal-labs/modal-client/blob/06c36d19661c1a91f444d2df523759eac6cc9272/modal/network_file_system.py#L212

Agree it's a bit confusing as is though.

(And to some extent, the CLI commands for NFS and Volumes, which work the same way)

Reply:
That makes sense, it would be confusing if the arguments were different. What do you think of the `add_file` / `add_dir` instead then? Might be clearer if it has a different name